docs/design/diagnostics.md

```diff
@@ -0,0 +1,293 @@
+  Severity() Severity
+  Summary()  string
+  Detail()   string
+
+  Equal(other Diagnostic) bool
+}
+
+type DiagnosticWithLogger interface {
+  Diagnostic
+
+  Log(ctx context.Context)
+}
+
+// This could be folded into Diagnostic
+type DiagnosticWithPath interface {
+  Diagnostic
+
+  Path() *tftypes.AttributePath // See also: https://github.com/hashicorp/terraform-plugin-framework/issues/81
+}
+
+type Diagnostics []Diagnostic
+
+func (d Diagnostics) Append(in ...Diagnostic)
+func (d Diagnostics) HasError() bool
+func (d Diagnostics) toTfprotov6Diagnostics() []*tfprotov6.Diagnostic
+```
+
+With an example implementation:
+
+```go
+// ErrorDiagnostic is a generic diagnostic with error severity.
+type ErrorDiagnostic struct {
+  Summary string
+  Detail  string
+}
+
+func (d ErrorDiagnostic) Severity() Severity { return SeverityError }
+func (d ErrorDiagnostic) Summary() string { return d.Summary }
+func (d ErrorDiagnostic) Detail() string { return d.Detail }
+func (d ErrorDiagnostic) Equal(other Diagnostic) bool {
+  other, ok := other.(ErrorDiagnostic)
+
+  if !ok {
+    return false
+  }
+
+  return other.Summary() == d.Summary() && other.Detail() == d.Detail()
+}
+
+// ValidationErrorDiagnostic represents an error during validation.
+type ValidationErrorDiagnostic struct {
+  ErrorDiagnostic
+}
+
+func (d ValidationErrorDiagnostic) Equal(other Diagnostic) bool {
+  other, ok := other.(ValidationErrorDiagnostic)
+
+  if !ok {
+    return false
+  }
+
+  return other.Summary() == d.Summary() && other.Detail() == d.Detail()
+}
+```
+
+It then becomes possible to distinguish between diagnostic types (in this case the "source" of a diagnostic) while still supporting additional functionality defined by the interface. These extended diagnostics can also implement other framework-defined capabilities, such as implementing the example `Log()` method in this situation.
+
+### Error Interface
+
+On top of modeling as an interface, diagnostics could be further marked as an extension of the `error` interface:
+
+```go
+package diag
+
+type Diagnostic interface {
+  error
+
+  DiagnosticSeverity() Severity
+  DiagnosticSummary()  string
+  DiagnosticDetail()   string
+}
+
+// This could be folded into Diagnostic
+type DiagnosticWithPath interface {
+  Diagnostic
+
+  DiagnosticPath() *tftypes.AttributePath // See also: https://github.com/hashicorp/terraform-plugin-framework/issues/81
+}
+```
+
+This would allow a single diagnostic to be usable in place of where an `error` would be returned and the type would pick up associated `errors` package capabilities. All implementations would be required for including an `Error()` method and optionally other `Unwrap()` etc. methods to satisfy other `errors` capabilities, which could introduce verbosity concerns. It also might be unclear what the semantics of a warning diagnostic are in terms of the `error` interface -- this implementation would treat all warning diagnostics as errors, which seems less than ideal.
```

Comment:
There's some prior art related to this in both HCL and in Terraform, which I wanted to note in case it's useful for thinking through the tradeoffs here. I'm not intending this as any recommendation to do anything in particular, because there are likely some different requirements and preferences at the SDK layer than at these two layers, which you all are in a far better place to mull than I am.

* HCL's `hcl.Diagnostics` implements `error` in a way similar to what's described here, but it's at the level of the group of diagnostics rather than at the level of a single diagnostic so that a group of diagnostics can return together as a single `error` value.

    While this worked okay for the internals of HCL where everything was generally in agreement about how diagnostics work, it ended up causing friction once we started using diagnostics in Terraform where there's more of a blend of native diagnostics handling and traditional `error` handling. In particular, we got caught out a few times by the fact that a `hcl.Diagnostics` which contains only `hcl.DiagWarning` diagnostics appears as a non-`nil` `error`, even though it's not actually describing an error condition, which I think is the same concern raised in the paragraph above.

* In response to those problems, we created [Terraform's own `tfdiags` package](https://pkg.go.dev/github.com/hashicorp/terraform/internal/tfdiags), which has [`tfdiags.Diagnostics`](https://pkg.go.dev/github.com/hashicorp/terraform/internal/tfdiags#Diagnostics) as a more general analog to `hcl.Diagnostics`. Part of that design was to intentionally make `tfdiags.Diagnostics` _not_ implement `error`, and instead it has a method [`Err`](https://pkg.go.dev/github.com/hashicorp/terraform/internal/tfdiags#Diagnostics.Err) which returns an `error` which is `nil` unless `diags.HasErrors()`, thus preserving the usual meaning of a non-`nil` error at the expense of then losing track of warnings that aren't accompanied by an error (because there's nowhere to put them in an `error` value).

    For some particularly gnarly cases we also have [`ErrWithWarnings`](https://pkg.go.dev/github.com/hashicorp/terraform/internal/tfdiags#Diagnostics.ErrWithWarnings) which essentially recovers the HCL approach of potentially returning a weird `error` that might actually only be reporting warnings. In the few cases where we use that the caller needs to be careful to check whether the `error` value has the type [`NonFatalError`](https://pkg.go.dev/github.com/hashicorp/terraform/internal/tfdiags#NonFatalError) and treat it as warnings only in that case.

* Tangentially, we also made [the `Append` method of `Diagnostics`](https://pkg.go.dev/github.com/hashicorp/terraform/internal/tfdiags#Diagnostics) able to accept naked `error` values and turn them into (often lower-quality) diagnostic errors. It also recognizes `error` values previously returned from `Diagnostics.Err` or `Diagnostics.NonFatalErr` and recovers the original diagnostics from them, which allows losslessly(-ish) sending diagnostics through a function that returns `error` because e.g. it needs to interact with other typical Go error-handling patterns.

None of these designs are ideal by any means, but the `tfdiags.Diagnostics` design has been working reasonably well and I personally think it was an improvement over the `hcl.Diagnostics` design in terms of it being easier to use it correctly.

Reply:
Thank you for this additional context, @apparentlymart! I'm wondering if it might be good to capture some of this related context into the design documentation in the prior art section for posterity. 🤔 

Given the various complexities mentioned, it makes it feel more pertinent to treat diagnostics as explicitly different than regular `error`s to reduce cognitive load. We don't have a use case for going from a diagnostic to an `error` currently, which probably simplifies our implementation. We do have the use case for promoting an `error` to a diagnostic (especially outside the framework since that is how vendor SDKs are generally written), but maybe requiring framework and provider developers work through converting them explicitly is a good nudge for writing better messaging rather than just writing the helper function to wholesale move the error string into an error diagnostic summary.